### PR TITLE
Quiz grading

### PR DESCRIPTION
**Pull request by blahajjessie**

### CODE REVIEW

File: frontend/pages/upload.tsx [hunk 1]
```diff
@@ -11,6 +15,8 @@ enum UploadStatus {
 	Done,
 }
 
+const inter = Inter({ subsets: ['latin'] });
+
```
Comment (190n) [inline frontend/pages/upload.tsx]:
```suggestion
```
Unused

File: frontend/pages/upload.tsx [hunk 2]
```diff
@@ -1,3 +1,7 @@
+import Image from 'next/image';
+import { Inter } from 'next/font/google';
+import MatchMediaWrapper from '../components/MatchMediaWrapper';
+
```
Comment (190n) [inline frontend/pages/upload.tsx]:
```suggestion
```

Unused/duplicates (really just revert all the changes to upload.tsx).

File: backend/src/game.ts [hunk 3]
```diff
@@ -116,21 +221,28 @@ export default function registerGameRoutes(app: Express) {
 		const index = parseInt(req.params.index);
 		const game = games.get(gameId);
 
+		let answer: number;
+		try {
+			answer = parseInt(req.body.answer);
+		} catch (e) {
+			answer = -1;
+		}
+
```
Comment (190n) [inline backend/src/game.ts]: Revert all the changes around checking `answer`. The way it's done on main is correct. See https://github.com/blahajjessie/kakaw/pull/7#discussion_r1176867389.

File: backend/src/game.ts [hunk 4]
```diff
@@ -33,26 +36,86 @@ export interface Quiz {
...
+	const game = games.get(gameId);
+	if (!game) return;
+	const users = getUsers(game);
+	const userSockets = connections.get(gameId);
+	if (userSockets === undefined) {
+		// Player List Not in Connections
+		return;
+	}
+	// TODO: only send question data (dont send answers, etc)
+	const question = JSON.stringify(game.quizData.questions[game.activeQuestion]);
```
Comment (190n) [inline backend/src/game.ts]: I don't think you need to stringify here. `sendMessage` does that.

File: backend/src/game.ts [hunk 5]
```diff
@@ -33,26 +36,86 @@ export interface Quiz {
...
+			].correctAnswers.includes(user.answers[game.activeQuestion]),
+
+			correctAnswers:
+				game.quizData.questions[game.activeQuestion].correctAnswers,
+
+			score: user.scores.reduce((a, b) => a + b, 0),
+			scoreChange: user.scores[game.activeQuestion],
+			time: user.times[game.activeQuestion],
+		};
+		let resp = JSON.stringify(endResp);
```
Comment (190n) [inline backend/src/game.ts]: See my review on beginQuestion. No need to stringify.

**Review by 190n — CHANGES_REQUESTED** [review]
Few more changes but nothing major. This is looking good.

There's a lot of repetition with stuff like checking if a game/question is valid, broadcasting messages to users, etc. We can merge this for now but I'm gonna think about how to refactor some of that stuff with less repetition.

File: backend/src/game.ts (hunk 4 above)
Comment (celdond) [reply backend/src/game.ts]: Second this, small error from my original code.

File: frontend/pages/upload.tsx (hunk 2 above)
Comment (celdond) [reply frontend/pages/upload.tsx]: Second reversion.  Out of scope of backend.

**Review by celdond — COMMENTED** [review]
No immediate revisions beyond what Ben pointed out already.
nextQuestion and beginQuestion could easily be merged, but we can discuss doing that at the beginning of the next sprint.
Good work.

File: backend/src/game.ts [hunk 6]
```diff
@@ -33,25 +36,86 @@ export interface Quiz {
...
+	const game = games.get(gameId);
+	if (!game) return;
+	const users = getUsers(game);
+	const userSockets = connections.get(gameId);
+	if (userSockets === undefined) {
+		// Player List Not in Connections
+		return;
+	}
+
+	users.forEach(function (value: UserId) {
```
Comment (190n) [inline backend/src/game.ts:67]: For future reference, you can do `for (const [id, user] of users)`. I'm not gonna make you change that before merging though.

File: backend/src/game.ts [hunk 7]
```diff
@@ -158,6 +264,7 @@ export default function registerGameRoutes(app: Express) {
 		}
 
 		res.status(200).send({ ok: true });
+		res.status(200).send({ ok: true });
```
Comment (190n) [inline backend/src/game.ts]:
```suggestion
```

duplicate

File: backend/src/game.ts [hunk 8]
```diff
@@ -183,8 +290,10 @@ export default function registerGameRoutes(app: Express) {
 		// 	games.set(gameId, playerSet);
 		// }
 
-		const username: UserId = body.username;
-		if (game.users.has(username)) {
+		// TODO This validation might be wrong
```
Comment (190n) [inline backend/src/game.ts]:
```suggestion
```

Looks fine to me

**Review by 190n — APPROVED** [review]

**Review by 190n — APPROVED** [review]

**Review by celdond — APPROVED** [review]

**Pull request merged**